In [1]:
#importing python libraries
import pandas as pd

In [2]:
#loading the dataset and displays first 10 observations
df = pd.read_excel(r"C:\Users\user\Documents\Data Analytics and Visualizations\students_dataset.xlsx")
df.head(10)

,stud_id,age_yrs,sex,course_taken,yr_of_study,att_perc,study_hrs_week,assgn_score,final_exam,fam_income_kes,transport_mode,phone_no,lecturer_notes,temp_var_delete
0,STU1000,23.0,Male,Statistics,1,79.5,21.6,50.4,75.7,38134.0,Boda,732008519,Good,36
1,STU1001,31.0,F,Statistics,3,94.8,22.2,46.3,64.4,47400.0,Bus,762821494,Late,97
2,STU1002,27.0,female,Computer Science,2,67.9,18.0,63.3,91.5,57427.0,bus,797929059,-,15
3,STU1003,24.0,male,Applied Maths,4,92.4,20.0,51.7,64.8,34452.0,Boda,795081844,Needs follow-up,62
4,STU1004,23.0,Male,Computer Science,4,47.1,21.1,66.7,70.7,32719.0,MATATU,768986770,Excellent,3
5,STU1005,35.0,female,Computer Science,4,86.3,7.1,86.5,82.0,42081.0,bus,731610424,Late,93
6,STU1006,27.0,Female,IT,2,100.0,6.5,43.2,54.3,44750.0,Boda,732532734,Excellent,91
7,STU1007,27.0,Male,Data Science,3,69.3,8.6,69.6,20.5,14015.0,Matatu,776804231,NaN,5
8,STU1008,20.0,female,Data Science,3,94.8,16.0,84.5,53.3,39436.0,Boda,786548260,Excellent,56
9,STU1009,24.0,Female,Economics,1,100.0,13.2,66.4,77.5,34990.0,Boda,723605386,Late,31


### Structure of the Dataset

In [3]:
#Dimensions of the dataset
df.shape

(1225, 14)

In [4]:
#Total number of observations
df.shape[0]

1225

In [5]:
#Total number of variables
df.shape[1]

14

In [6]:
#Names of variables
df.columns

Index(['stud_id', 'age_yrs', 'sex', 'course_taken', 'yr_of_study', 'att_perc',
       'study_hrs_week', 'assgn_score', 'final_exam', 'fam_income_kes',
       'transport_mode', 'phone_no', 'lecturer_notes', 'temp_var_delete'],
      dtype='str')

In [7]:
#Data type of all variables
df.dtypes

stud_id                str
age_yrs            float64
sex                    str
course_taken           str
yr_of_study          int64
att_perc           float64
study_hrs_week     float64
assgn_score        float64
final_exam         float64
fam_income_kes     float64
transport_mode         str
phone_no             int64
lecturer_notes         str
temp_var_delete      int64
dtype: object

#### Structure of the dataset
- The dataset has 14 variables and 1225 observations.
- The variables have two data types: string and float

### Missing Values

In [8]:
#Checking if there are any missing values in the dataset
df.isnull().values.any()

np.True_

- There are missing values in the dataset

In [9]:
df.isnull()

,stud_id,age_yrs,sex,course_taken,yr_of_study,att_perc,study_hrs_week,assgn_score,final_exam,fam_income_kes,transport_mode,phone_no,lecturer_notes,temp_var_delete
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1220,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1221,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1222,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1223,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [10]:
#Total number of missing values in the dataset
df.isnull().sum().sum()

np.int64(316)

- The entire dataset has 316 missing values

In [11]:
#Total number of missing values in each variable
df.isnull().sum()

stud_id              8
age_yrs             25
sex                  0
course_taken         0
yr_of_study          0
att_perc             0
study_hrs_week       0
assgn_score         27
final_exam          25
fam_income_kes      25
transport_mode       0
phone_no             0
lecturer_notes     206
temp_var_delete      0
dtype: int64

#### Filling the missing values
- The 8 missing values in the stud_id column can't be filled since they are a unique identifier

#### 1. Using the measures of central tendency
a) Filling missing values in final_exam variable using the mean since preserves the overall average of the dataset.

In [12]:
df.final_exam = df.final_exam.fillna(df.final_exam.mean())

In [13]:
#checking if the missing values in the final_exam variable have been filled
df.final_exam.isnull().sum()

np.int64(0)

b) Filling missing values in fam_income_kes variable using median because median represents the "typical" family income without being pulled by extremely high or low values, i.e. outliers since income data is typically slewed.

In [14]:
df.fam_income_kes = df.fam_income_kes.fillna(df.final_exam.mean())

In [15]:
#checking if the missing values in the fam_income_kes variable have been filled
df.fam_income_kes.isnull().sum()

np.int64(0)

c) Filling missing values in lecturer_notes variable using mode (the most occuring value) because it is a categorical variable.

In [16]:
#the mode of the variable lecturer_notes
df.lecturer_notes.mode()

0    Good
Name: lecturer_notes, dtype: str

In [17]:
#Using the mode to fill the missing values of the variable lecturer_notes
df.lecturer_notes = df.lecturer_notes.fillna('Good')

In [18]:
#checking if the missing values in the lecturer_notes variable have been filled
df.lecturer_notes.isnull().sum()

np.int64(0)

#### 2. Using the Forward Fill (ffill) method to fill missing values in age_yrs variable because students with nearby IDs are likely have similar ages.

In [19]:
df.age_yrs = df.age_yrs.ffill()

In [20]:
#checking if the missing values in the lecturer_notes variable have been filled
df.age_yrs.isnull().sum()

np.int64(0)

#### 3. Using the Backward Fill (bfill) method to fill missing values in assgn_score because It's a good placeholder as it shows an actual score from the dataset.

In [21]:
df.assgn_score = df.assgn_score.ffill()

In [22]:
#checking if the missing values in the assgn_score variable have been filled
df.assgn_score.isnull().sum()

np.int64(0)

#### Difference between forward fill and backward fill methods
- The difference is in the direction in which known valid observations are propagated to fill in missing data.
- Forward fill method fills in missing values by propagating the last known valid observation forward to the next missing one.
- Backward fill method fills in missing values by using the next known valid observation backward to the previous missing one.

In [23]:
#Checking if all missing values have been filled except stud_id
df.isnull().sum()

stud_id            8
age_yrs            0
sex                0
course_taken       0
yr_of_study        0
att_perc           0
study_hrs_week     0
assgn_score        0
final_exam         0
fam_income_kes     0
transport_mode     0
phone_no           0
lecturer_notes     0
temp_var_delete    0
dtype: int64

In [24]:
#Displaying the dataset after handling missing values
df.isnull()

,stud_id,age_yrs,sex,course_taken,yr_of_study,att_perc,study_hrs_week,assgn_score,final_exam,fam_income_kes,transport_mode,phone_no,lecturer_notes,temp_var_delete
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1220,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1221,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1222,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1223,False,False,False,False,False,False,False,False,False,False,False,False,False,False


### Deleting variables

In [25]:
#Deleting the stud_id variable from the dataset
df.drop('stud_id', axis = 1, inplace = True)

In [26]:
df.columns

Index(['age_yrs', 'sex', 'course_taken', 'yr_of_study', 'att_perc',
       'study_hrs_week', 'assgn_score', 'final_exam', 'fam_income_kes',
       'transport_mode', 'phone_no', 'lecturer_notes', 'temp_var_delete'],
      dtype='str')

- The student id variable is not useful for statistical analysis because it used for identification only and it isn't measurable and therefore cannot be used for analysis

### Duplicate records

In [27]:
#checking whether duplicate observations exist in the dataset
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1220     True
1221     True
1222     True
1223     True
1224    False
Length: 1225, dtype: bool

In [28]:
#checking whether duplicate observations exist in the dataset 
df.duplicated().any()

np.True_

In [29]:
#Total number of duplicates observations
df.duplicated().sum()

np.int64(21)

In [30]:
#Display all duplicate observations in the dataset
df[df.duplicated()]

,age_yrs,sex,course_taken,yr_of_study,att_perc,study_hrs_week,assgn_score,final_exam,fam_income_kes,transport_mode,phone_no,lecturer_notes,temp_var_delete
1200,35.0,Male,IT,2,76.2,22.4,61.1,61.3,34427.0,Bus,788549191,-,3
1201,33.0,Male,Economics,3,65.3,10.1,50.4,61.0,46108.0,Boda,712644688,Needs follow-up,10
1205,19.0,F,Data Science,2,92.7,17.4,61.2,57.8,72894.0,Bus,794866532,Needs follow-up,25
1206,34.0,F,Statistics,4,83.2,6.3,79.2,64.4,53179.0,Walk,729221071,Good,52
1207,18.0,M,Statistics,2,74.7,18.0,100.0,63.7,29156.0,Boda,792529753,Excellent,12
1208,17.0,M,Statistics,1,82.5,11.2,89.9,49.4,47816.0,Private,714522847,-,58
1209,24.0,female,IT,3,85.0,23.0,79.6,82.2,21176.0,bus,716367173,-,17
1210,17.0,Female,Computer Science,2,73.2,17.0,80.5,36.3,4477.0,Boda,788427412,Good,97
1211,27.0,Male,Economics,2,58.7,25.8,54.5,43.0,-7407.0,Walk,731656992,Late,5
1212,29.0,Male,Statistics,4,98.5,8.4,63.8,64.3,57109.0,bus,786969534,Good,53


In [31]:
#Remove all duplicate observations
df = df.drop_duplicates()

In [32]:
#Confirming that duplicate observations no longer exist in the cleaned dataset
df.duplicated().sum()

np.int64(0)

### Standardization

In [33]:
#Categories in the sex variavle
df.sex.value_counts()

sex
female    215
F         212
Female    208
Male      197
male      196
M         176
Name: count, dtype: int64

- There are inconsistencies in the categories since both the male and female students are identified using different variations instead of just one identification, i.e, for male students there is Male, male, M and for female students there is Female, female and F.

In [34]:
#Standardizing the format of the sex variable
df.sex = df.sex.replace({'F':'Female', 'female': 'Female', 'male':'Male', 'M': 'Male'})

In [35]:
#verifying the standardization was successful
df.sex.value_counts()

sex
Female    635
Male      569
Name: count, dtype: int64

### Saving the cleaned version of the dataset as a new file

In [36]:
df.to_excel('cleaned_student_dataset.xlsx')

#### Why is it important to save the cleaned dataset separately from the raw dataset.
- It allows you to load it directly in subsequent analysis without reprocessing everything saving time.

#### Importance of data cleaning and preprocessing before conducting statistical analysis
- Improves data accuracy: Removes errors and duplicate records that can distort statistical measures like means, variances, and correlations.
- Handles Missing Information: Prevents misleading interpretations since missing values can significantly affect analysis outcomes.
- Reduces noise and inconsistency to improve the detection of meaningful trends and relationships.
- Enhancing data integration: when combining data from multiple sources, preprocessing helps create a unified and coherent dataset.